README.md
# Leap for PHP

Leap is an database management library for the [Kohana PHP Framework](http://kohanaframework.org) and is designed to work with DB2,
Drizzle, Firebird, MariaDB, MS SQL, MySQL, Oracle, PostgreSQL, and SQLite.  Leap acts as a common interface between the different
database dialects and connections.  It provides a powerful query builder and ORM.  Leap's ORM is based on the active record design
pattern, which utilizes PHP objects to model database tables.

Make sure to "Star" this project if you like it.

## Motivation

Leap is meant to be a simple, clean project.  The primary goal of Leap is to create an ORM for the Kohana PHP Framework that works
with all major databases and is meant to be a true Kohana module so that it could be just dropped into Kohana's modules folder and
work.  Even though other ORMs like [Doctrine](http://www.doctrine-project.org/projects/orm) are adaptable to Kohana, doing so
requires a lot of work.  Another goal for the development of Leap is to create an ORM for Kohana that can harness the power of
composite keys, which many other ORMs (e.g. [Kohana's official ORM](https://github.com/kohana/orm), [Jelly](https://github.com/creatoro/jelly),
and [Sprig](https://github.com/sittercity/sprig/)) cannot handle.

## Features

Leap provides a number of features, such as:

* Plugins for DB2, Drizzle, Firebird, MariaDB, MS SQL, MySQL, Oracle, PostgreSQL, and SQLite.
* Designed to work in conjunction with other database tools for Kohana.
* [Config file for designating the database driver (e.g. PDO) and connection strings](http://orm.spadefootcode.com/tutorials/setting-up-a-database-connection/).
* Classes are easily extensible.
* A [database connection pool](http://orm.spadefootcode.com/tutorials/establishing-a-database-connection/) for managing resources.
* A powerful [query builder for creating SQL statements](http://orm.spadefootcode.com/tutorials/building-sql-statements/).
* Sanitizes data to help prevent SQL injection attacks.
* Capable of handling non-integers primary keys.
* Supports composite primary keys and composite foreign keys.
* Enforces strong data types on [database fields](http://orm.spadefootcode.com/tutorials/mapping-a-model/#fields).
* Allows [field aliases](http://orm.spadefootcode.com/tutorials/mapping-a-model/#aliases) to be declared.
* Makes working with certain database fields easy with [field adaptors](http://orm.spadefootcode.com/tutorials/mapping-a-model/#adaptors).
* A set of Auth classes for authenticating user logins.
* A toolkit of useful functions.
* [Leap's API](http://orm.spadefootcode.com/api/annotated.html) that documents each class.
* Lots of [tutorials](http://orm.spadefootcode.com/tutorials/index/).

## Getting Started

To start using Leap, follow these steps:

1. Just download the module (see below regarding as to which branch) from github.
2. Unzip the download to the modules folder in Kohana.
3. Rename the uncompressed folder to "leap".
4. Modify leap/config/database.php.
5. Add "leap" as a module to application/bootstrap.php.
6. Begin creating your models in the application/classes/model/leap/ folder.

For more information, see the tutorial on [installing Leap](http://orm.spadefootcode.com/tutorials/installing-leap/).

### About Branches

* 3.3/master  - PHP 5.3+, Kohana 3.3.X, maintained, stable
* 3.3/develop - PHP 5.3+, Kohana 3.3.X, maintained, unstable
* 3.3/legacy  - PHP 5.2+, Kohana 3.3.X, deprecated, stable
* 3.2/master  - PHP 5.3+, Kohana 3.2.X, maintained, stable
* 3.2/develop - PHP 5.3+, Kohana 3.2.X, maintained, unstable
* 3.2/legacy  - PHP 5.2+, Kohana 3.2.X, deprecated, stable
* 3.1/master  - PHP 5.3+, Kohana 3.1.X, maintained, stable
* 3.1/develop - PHP 5.3+, Kohana 3.1.X, maintained, unstable
* 3.1/legacy  - PHP 5.2+, Kohana 3.1.X, deprecated, stable

## Required Files

The Leap ORM module is meant to be completely independent of other Kohana modules.  However, it is recommended that Kohana's database module be
installed as well so that you can utilize the \Database\Expression class.  As for the files within Leap, you can remove any database plugins that
you are not using.  This is possible because each database plugin in Leap is considered independent of the others.

## Documentation

This project is accompanied by [a companion Web site](http://orm.spadefootcode.com), which documents the [API for the Leap ORM](http://orm.spadefootcode.com/api/annotated.html)
and has a number of [examples and tutorials](http://orm.spadefootcode.com/tutorials/index/). You can also find other tutorials and examples
online (please let us know if you find one that we should highlight here).

## Further Assistance

Although Leap is simple to use with the Kohana PHP Framework, you can get further assistance by asking questions on either [Kohana's Forum](http://forum.kohanaframework.org/)
or [Stack Overflow](http://stackoverlow.com). You can also send an email to spadefoot.oss@gmail.com.

## Reporting Bugs & Making Recommendations

If you find a bug in the code or if you would like to make a recommendation, we would be happy to hear from you.  Here are three methods
you can use to notify us:

* Log an issue in this project's [issue tracker](https://github.com/spadefoot/kohana-orm-leap/issues?sort=comments&direction=desc&state=open).
* Create a fork of this project and submit a [pull request](http://help.github.com/send-pull-requests/).
* Send an email to spadefoot.oss@gmail.com.

## Known Issues

Please see this project's [issue tracker](https://github.com/spadefoot/kohana-orm-leap/issues?sort=comments&direction=desc&state=open) on github for any known issues.

## Updates

Make sure that you add yourself as a watcher of this project so that you can watch for updates.  If you would like to be notified directly
via email please send an email to spadefoot.oss@gmail.com.

## Future Development

This project is constantly being improved and extended. If you would like to contribute to Leap, please fork this project and then send
us your additions/modifications using a [pull request](http://help.github.com/send-pull-requests/).

## License

### Apache v2.0

Copyright © 2011–2014 Spadefoot Team.

Unless otherwise noted, Leap is licensed under the Apache License, Version 2.0 (the "License"); you may not use these files except in
compliance with the License. You may obtain a copy of the License at:

[http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
and limitations under the License.


ChangeLog.txt
Change Log for Leap ORM

2013-01-28
* Changed Base\DB\ORM\Model::data_source() to Base\DB\ORM\Model::data_source($instance = 0).
* Changed all instance variables $source to $data_source.

2013-01-01
* Updated copyright information.
* Removed ?> at the end of every PHP file.

2012-12-29
* Refactored constructors for reader classes.

2012-12-05
* Added @override property in inline documentation.

2012-12-04
* Refactored connection drivers to incorporate the SQL data reader paradigm (Issue #72 & #73).
* Moved the base connection driver class to a subfolder.

2012-11-28
* Added role support to the Firebird driver classes (Issue #74)
* Reverting a bad refactoring to the Base\DB\Oracle\Connection\PDO::open() (Issue #74). 

2012-11-14
* Updated 3.3 branch with 3.2 changes.
* Moved and renamed exception classes.
* Moved and renamed object class.

2012-10-15
* Adding new Core\Data\BitField class.
* Redid the Base\DB\ORM\Field\Bit class.
* Changing Core\Data\ByteString::BYTE_DATA back to Core\Data\ByteString::BINARY_DATA.
* Added the ability to format strings in Base\Core\Data\ByteString.
* Minor inline documentation changes.

2012-08-24
* Added a new 'count' method to SQL select builder.
* Added some sample MPTT SQL schema files.
* Updated MPTT methods.
* Corrected some inline documentation.
* Replaced some string operators with constants.
* Replaced is_null function with === comparisons.
* Added a new 'regex' property to 'string' field.
* Replaced 'self' keywords, get_called_class(), and call_user_func() with PHP 5.3+ equivalents.
* Added some new UOM conversions to config.
* Replace && and || with AND and OR, respectfully.
* Added adaptors and aliases to Base\DB\ORM\Model::as_array().
* Added create and update methods to Base\DB\ORM\Model.
* Replace spl_object_hash() with Core\Object::__hashCode().
* Base\DB\ORM\Field\Blob now uses the Core\Data\ByteString class.
* Renamed DB\ORM\Field\Adaptor\Array to DB\ORM\Field\Adaptor\ArrayList.
* Renamed some variables.

2012-08-14
* Reworked validation in field classes (Issues #50, #59, #60).

2012-06-14
* Changing Core\Data\ByteString::BINARY_DATA to Core\Data\ByteString::BYTE_DATA.

2012-06-06
* Merging pull request #41 as per nyunyu95's request.

2012-05-31
* Fixed an issue with the connection pool, where the connection was not being openned.

2012-05-29
* Changed "foreach" loop to a "do/while" loop in Base\DB\Connection\Pool.
* Merged pull requests #38 and #39 per nexeck's request.

2012-05-25
* Reworked the internal workings of the connection pool, adding the ability to create new connections and to limit the number of connections.
* Removed some unneeded, commented code.

2012-05-20
* Renamed some variables to better describe their usage in connection classes.
* Changed how "type" property is handled in data source.

2012-05-15
* Fixed bug in Base\DB\ORM\Field\Decimal::validate().
* Cleaned-up and improved Base\DB\ORM\Model::hash_code() (Issue #25 & #26).
* Minor improvements to Base\Core\Data\ByteString class.
* Updated inline documentation.

2012-05-11
* Fixing a bug with like statement's escape clause.

2012-05-10
* Added a new class to handle binary data.
* Renamed Base\DB\DataSource::setup($settings, $id = NULL) to Base\DB\DataSource::init($settings, $id = NULL).
* Made "dialect" and "type" synonymous in Base\DB\DataSource so that "type" can be defined differently (e.g. PDO) in the connection config for Kohana's ORM.
* Fixed a bug with Leap's Auth login/logout when User Tokens were manually removed from database, merging CubedEye's pull request (Issue #35).

2012-04-09
* Added new table schema files for MS SQL and Oracle.
* Made a few minor changes to some of the other schema files.

2012-04-08
* Made base auth classes abstract.
* Added new table schema files for Drizzle and MariaDB.
* Simplified error handling in database connection classes.
* Added a connection check in quote methods and get_last_insert_id methods.
* Added the ability to set a database's character set pre matesko's request (Issue #24).
* Changed the default value of Base\DB\DataSource::$charset.

2012-04-07
* Added new table schema files for PostgreSQL and SQLite.

2012-04-05
* Fixed misspelling in model class pre Samnan's request (Issue #32).
* Updated auth models.
* Reworked table schema files for mysql.

2012-04-03
* Standardized column names in auth tables/classes to make migrating over to Leap easier (Issue #31).
* Moved table schema files into subfolders.
* Moved SQL dialect config files into a subfolder.

2012-04-01
* Added a new "Has Many Through" relation (Issues #22).

2012-03-31
* Moved "auth_table.sql" to a new folder and renamed it to "auth-mysql.sql".
* Fixed a bug with the method signature for Base\DB\ORM\Proxy::column().
* Commented class methods in all auth classes.

2012-03-27
* Merged pull request from CubedEye (Issues #10 & #27)

2012-03-22
* Added new Base\DB\ORM\Model::is_saved() method.

2012-03-08
* Fixed a bug in Base\DB\ORM\Model::hash_code() per matesko's request (Issue #25).
* Fixed a bug in Base\DB\ORM\Model::load() per matesko's request (Issue #25).

2012-03-05
* Models can now generate HTML form labels and controls using field data.

2012-02-22
* Added all function to Base\DB\SQL\Select\Builder, Base\DB\SQL\Select\Proxy, and Base\DB\ORM\Select\Proxy.
* Added prepare_wildcard($expr) to expression classes.
* DB\ORM now adds table prefix to wildcard to avoid problems when joining.
* Updated inline documentation.

2012-02-10
* Changed the method signature to Base\DB\SQL\Expression_Interface::prepare_operator() and updated implementing classes.
* Updated some inline documentation.

2012-02-09
* Fixed the missing parentheses in Base\DB\ORM\Field\Double.
* Redid the quote function in connection classes.

2012-02-07
* Fixed a bug in how Oracle escapes strings per ekarlso's request (Issue #18).
* Improved how strings are escaped.
* Renamed Base\DB\Connection\Driver::escape_string($string) to Base\DB\Connection\Driver::quote($string, $like = FALSE).

Leap/Core/Model/User.php
<?php

/**
 * Copyright © 2011–2014 Spadefoot Team.
 * Copyright © 2012 CubedEye.
 *
 * Unless otherwise noted, Leap is licensed under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License
 * at:
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Leap\Core\Model {

	use \Leap\Core\DB;

	/**
	 * This class represents a record in the "users" table.
	 *
	 * @access public
	 * @class
	 * @package Leap\Core\Model
	 * @version 2014-01-25
	 */
	class User extends DB\ORM\Model {

		/**
		 * This constructor instantiates this class.
		 *
		 * @access public
		 */
		public function __construct() {
			parent::__construct();

			$this->fields = array(
				'id' => new DB\ORM\Field\Integer($this, array(
					'max_length' => 11,
					'nullable' => FALSE,
					'unsigned' => TRUE,
				)),
				'email' => new DB\ORM\Field\String($this, array(
					'max_length' => 254,
					'nullable' => FALSE,
				)),
				'username' => new DB\ORM\Field\String($this, array(
					'default' => '',
					'max_length' => 32,
					'nullable' => FALSE,
				)),
				'password' => new DB\ORM\Field\String($this, array(
					'max_length' => 64,
					'nullable' => FALSE,
				)),
				// Personal Details
				'firstname' => new DB\ORM\Field\String($this, array(
					'default' => NULL,
					'max_length' => 35,
					'nullable' => TRUE,
				)),
				'lastname' => new DB\ORM\Field\String($this, array(
					'default' => NULL,
					'max_length' => 50,
					'nullable' => TRUE,
				)),
				// Account Status Details
				'activated' => new DB\ORM\Field\Boolean($this, array(
					'default' => TRUE,
					'nullable' => FALSE,
				)),
				'banned' => new DB\ORM\Field\Boolean($this, array(
					'default' => FALSE,
					'nullable' => FALSE,
				)),
				'ban_reason' => new DB\ORM\Field\String($this, array(
					'default' => NULL,
					'max_length' => 255,
					'nullable' => TRUE,
				)),
				// Account Utility Details
				'new_password_key' => new DB\ORM\Field\String($this, array(
					'default' => NULL,
					'max_length' => 64,
					'nullable' => TRUE,
				)),
				'new_password_requested' => new DB\ORM\Field\Integer($this, array(
					'default' => NULL,
					'max_length' => 11,
					'nullable' => TRUE,
				)),
				'new_email' => new DB\ORM\Field\String($this, array(
					'default' => NULL,
					'max_length' => 254,
					'nullable' => TRUE,
				)),
				'new_email_key' => new DB\ORM\Field\String($this, array(
					'default' => NULL,
					'max_length' => 64,
					'nullable' => TRUE,
				)),
				// Account Metrics Details
				'logins' => new DB\ORM\Field\Integer($this, array(
					'default' => 0,
					'max_length' => 10,
					'nullable' => FALSE,
					'unsigned' => TRUE,
				)),
				'last_login' => new DB\ORM\Field\Integer($this, array(
					'default' => NULL,
					'max_length' => 10,
					'nullable' => TRUE,
					'unsigned' => TRUE,
				)),
				'last_ip' => new DB\ORM\Field\String($this, array(
					'default' => NULL,
					'max_length' => 39,
					'nullable' => TRUE,
				)),
			);

			$this->adaptors = array(
				'last_login_formatted' => new DB\ORM\Field\Adaptor\DateTime($this, array(
					'field' => 'last_login',
				)),
				'new_password_requested_formatted' => new DB\ORM\Field\Adaptor\DateTime($this, array(
					'field' => 'new_password_requested',
				)),
			);

			$this->relations = array(
				'user_roles' => new DB\ORM\Relation\HasMany($this, array(
					'child_key' => array('user_id'),
					'child_model' => 'User_Role',
					'parent_key' => array('id'),
				)),
				'user_token' => new DB\ORM\Relation\HasMany($this, array(
					'child_key' => array('user_id'),
					'child_model' => 'User_Token',
					'parent_key' => array('id'),
				)),
			);
		}

		/**
		 * This method completes the user's login.
		 *
		 * @access public
		 */
		public function complete_login() {
			$this->logins++;
			$this->last_login = time();
			$this->last_ip = \Request::$client_ip;
			$this->save();
		}

		/**
		 * This method returns the data source name.
		 *
		 * @access public
		 * @override
		 * @static
		 * @param integer $instance                     the data source instance to be used (e.g.
		 *                                              0 = master, 1 = slave, 2 = slave, etc.)
		 * @return string                               the data source name
		 */
		public static function data_source($instance = 0) {
			return 'default';	
		}

		/**
		 * This method returns the primary key for the database table.
		 *
		 * @access public
		 * @override
		 * @static
		 * @return array                                the primary key
		 */
		public static function primary_key() {
			return array('id');	
		}

		/**
		 * This method returns the database table's name.
		 *
		 * @access public
		 * @override
		 * @static
		 * @return string                               the database table's name
		 */
		public static function table() {
			return 'users';	
		}

	}

}